src/Microsoft.AspNetCore.Routing/Tree/TreeRouteBuilder.cs

```diff
@@ -392,5 +398,36 @@ private void AddEntryToTree(UrlMatchingTree tree, InboundRouteEntry entry)
                 return result == 0 ? x.Entry.RouteTemplate.TemplateText.CompareTo(y.Entry.RouteTemplate.TemplateText) : result;
             });
         }
+
+        private static bool RemainingSegmentsAreOptional(IList<TemplateSegment> segments, int currentParameterIndex)
+        {
+            for (var i = currentParameterIndex; i < segments.Count; i++)
+            {
+                if (!segments[i].IsSimple)
+                {
+                    // /{complex}-{segment}
+                    return false;
+                }
+
+                if (!segments[i].Parts[0].IsParameter)
+                {
+                    // /literal
+                    return false;
+                }
+
+                if (!IsOptionalOrCatchAllOrHasDefaultValue(segments[i].Parts[0]))
+                {
+                    // /{parameter}
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static bool IsOptionalOrCatchAllOrHasDefaultValue(TemplatePart part)
```

Comment:
This could just be merged into the above. I know you stylistically **love** long method names, but this isn't abstracting anything that a local variable doesn't do.